feature/s3/transfermanager/putobject.go

```diff
@@ -0,0 +1,772 @@
+	// was uploaded as a single PutObject call.
+	UploadID string
+
+	// The list of parts that were uploaded and their checksums. Will be empty
+	// if multipart upload was not used, and the object was uploaded as a
+	// single PutObject call.
+	CompletedParts []types.CompletedPart
+
+	// Indicates whether the uploaded object uses an S3 Bucket Key for server-side
+	// encryption with Amazon Web Services KMS (SSE-KMS).
+	BucketKeyEnabled bool
+
+	// The base64-encoded, 32-bit CRC32 checksum of the object.
+	ChecksumCRC32 *string
+
+	// The base64-encoded, 32-bit CRC32C checksum of the object.
+	ChecksumCRC32C *string
+
+	// The base64-encoded, 160-bit SHA-1 digest of the object.
+	ChecksumSHA1 *string
+
+	// The base64-encoded, 256-bit SHA-256 digest of the object.
+	ChecksumSHA256 *string
+
+	// Entity tag for the uploaded object.
+	ETag *string
+
+	// If the object expiration is configured, this will contain the expiration date
+	// (expiry-date) and rule ID (rule-id). The value of rule-id is URL encoded.
+	Expiration *string
+
+	// The bucket where the newly created object is put
+	Bucket *string
+
+	// The object key of the newly created object.
+	Key *string
+
+	// If present, indicates that the requester was successfully charged for the
+	// request.
+	RequestCharged types.RequestCharged
+
+	// If present, specifies the ID of the Amazon Web Services Key Management Service
+	// (Amazon Web Services KMS) symmetric customer managed customer master key (CMK)
+	// that was used for the object.
+	SSEKMSKeyId *string
+
+	// If you specified server-side encryption either with an Amazon S3-managed
+	// encryption key or an Amazon Web Services KMS customer master key (CMK) in your
+	// initiate multipart upload request, the response includes this header. It
+	// confirms the encryption algorithm that Amazon S3 used to encrypt the object.
+	ServerSideEncryption types.ServerSideEncryption
+
+	// The version of the object that was uploaded. Will only be populated if
+	// the S3 Bucket is versioned. If the bucket is not versioned this field
+	// will not be set.
+	VersionID *string
+}
+
+func (c Client) Upload(ctx context.Context, input *PutObjectInput, opts ...func(*Options)) (*PutObjectOutput, error) {
+	i := uploader{in: input, cfg: c, ctx: ctx}
+	// Copy ClientOptions
+	clientOptions := make([]func(*s3.Options), 0, len(c.options.PutClientOptions)+1)
+	clientOptions = append(clientOptions, func(o *s3.Options) {
+		o.APIOptions = append(o.APIOptions,
+			middleware.AddSDKAgentKey(middleware.FeatureMetadata, userAgentKey),
+			addFeatureUserAgent, // yes, there are two of these
+			func(s *smithymiddleware.Stack) error {
+				return s.Finalize.Insert(&setS3ExpressDefaultChecksum{}, "ResolveEndpointV2", smithymiddleware.After)
+			},
+		)
+	})
+	clientOptions = append(clientOptions, c.options.PutClientOptions...)
+	i.cfg.options.PutClientOptions = clientOptions
+	for _, opt := range opts {
+		opt(&i.cfg.options)
```

Comment:
You're mutating options here - you have to create a `Copy()` of them local to each client API and then apply functional options therein so the original is preserved.